pkg/utils/versions.go

```diff
@@ -0,0 +1,96 @@
+package utils
+
+import (
+	"encoding/xml"
+	"fmt"
+	"github.com/Masterminds/semver/v3"
+	"github.com/hashicorp/go-retryablehttp"
+	"net/url"
+	"sort"
+	"strings"
+)
+
+const (
+	repositoryUrl = "https://armory-cli-releases.s3.amazonaws.com/"
+
+	cliPrefix = "cli"
+)
+
+var client = retryablehttp.NewClient()
+
+// GetAllVersions returns versions available in S3 repository, sorted as SemVers in descending order.
+func GetAllVersions() ([]string, error) {
+	var versions []*semver.Version
+	more := true
+	params := url.Values{}
+	params.Set("list-type", "2")
+	params.Set("prefix", cliPrefix)
+	for more {
+		u := buildUrl(params)
+		httpRes, err := client.Get(u)
+		if err != nil {
+			return []string{}, err
```

Comment:
I would say that returning `[]string{}` is slightly surprising in go, though I would love to get into a big argument about it 😄 

- in general if there is an error, I return the zero-value of the non-error types. The zero value of a slice is `nil`, not an empty slice
- a `nil` slice has almost exactly the same properties as an empty slice. its `len` is zero, you can `range` over it without panicking, etc. `json.Marshal` turns it into `null`, not `[]`, which is not relevant here.

Reply:
definitely! totally forgot about it